File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/PasskeyEnrolment.kt
package com.loonext.android.features.settings

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import androidx.credentials.CreatePublicKeyCredentialRequest
import androidx.credentials.CreatePublicKeyCredentialResponse
import androidx.credentials.CredentialManager
import androidx.credentials.exceptions.CreateCredentialCancellationException
import androidx.credentials.exceptions.CreateCredentialException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.OkHttpClient
import okhttp3.Request

/**
 * #473 — enrolling a passkey on Android, through Credential Manager.
 *
 * ## What this file is, and is not
 *
 * It is a courier. GoTrue produces `PublicKeyCredentialCreationOptionsJSON`,
 * Credential Manager consumes it and produces `RegistrationResponseJSON`, and
 * GoTrue consumes that. Both are the shapes the WebAuthn spec defines, so the
 * two ends already agree and nothing here parses, re-encodes, or base64s
 * anything. No private key is ever visible to this process — the credential is
 * created and held by the platform, behind the user's own screen lock.
 *
 * ## Why the failures are told apart
 *
 * Three things go wrong here and they need three different answers:
 *
 * - **Dismissed.** Somebody pressed back on the sheet. Not an error; say
 *   nothing and leave the card as it was. Shouting at somebody who changed
 *   their mind teaches them the button is dangerous.
 * - **The domain is not associated with this app.** Until
 *   `https://<rpId>/.well-known/assetlinks.json` names this package and its
 *   signing certificate, every attempt fails, and it fails identically for the
 *   first person to try it and for a genuine bug. This one has to be named,
 *   because "try again" is advice that cannot work.
 * - **Anything else.** The platform's own message beats ours: "the operation
 *   either timed out or was not allowed" is what somebody needs to read when
 *   their fingerprint was not recognised.
 *
 * ## The relying party
 *
 * [WEBAUTHN_RP_ID] is the web app's host, not the API's. A passkey is scoped to
 * the RP that created it, so enrolling under a different id here would produce a
 * credential the web app could never use — the same account, two second factors
 * that each work in exactly one place.
 */

/**
 * The relying party every client enrols under. The web app enrols with
 * `window.location.hostname`, which is this, and the phones must match it or the
 * credentials do not travel.
 */
const val WEBAUTHN_RP_ID = "app.loonext.com"

/**
 * Whether this domain has authorised this app to hold passkeys for it.
 *
 * ## Why ask at all, rather than just offering the button
 *
 * Because until `/.well-known/assetlinks.json` names this package, EVERY
 * ceremony fails — and it fails the same way a genuine bug does. Offering a
 * button that cannot work is the one thing worse than not offering it, and this
 * product's audience is standing in somebody's driveway when they try it.
 *
 * ## Why a probe rather than a build flag
 *
 * A flag would need an app update to flip, and an app update takes weeks to
 * reach a phone. The association file is served by the web app, which deploys
 * in minutes. Reading the switch from the side that can move means passkeys
 * appear the day they start working, on handsets that were installed months
 * before.
 *
 * ## What it does NOT verify
 *
 * The signing fingerprint. Checking that this build's certificate is among the
 * ones listed would mean hashing our own APK signature — worth it if this were
 * a security decision, and it is not: the platform re-checks everything anyway
 * and refuses on any mismatch. This answers the weaker, sufficient question,
 * "has anybody configured this domain for Android at all", whose honest answer
 * today is no.
 *
 * Failure is `false`: no network, no passkey offer, and the authenticator app
 * is right there.
 */
suspend fun isPasskeyDomainAssociated(
    http: OkHttpClient,
    rpId: String = WEBAUTHN_RP_ID,
): Boolean = withContext(Dispatchers.IO) {
    try {
        val request = Request.Builder()
            .url("https://$rpId/.well-known/assetlinks.json")
            .build()
        http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return@withContext false
            val statements = Json.parseToJsonElement(
                response.body.string(),
            ) as? JsonArray ?: return@withContext false
            statements.any { entry ->
                val statement = entry as? JsonObject ?: return@any false
                val target = statement["target"]?.jsonObject
                val relations = statement["relation"] as? JsonArray
                target?.get("namespace")?.jsonPrimitive?.content == "android_app" &&
                    target["package_name"]?.jsonPrimitive?.content == ANDROID_PACKAGE &&
                    relations?.any {
                        it.jsonPrimitive.content == LOGIN_CREDS_RELATION
                    } == true
            }
        }
    } catch (_: Exception) {
        // Any failure — offline, DNS, malformed JSON — means we cannot show a
        // passkey option we can stand behind. The authenticator app path is
        // untouched and needs no network to be offered.
        false
    }
}

/** This app's package, as the statement file must name it. */
private const val ANDROID_PACKAGE = "com.loonext.android"

/** The relation WebAuthn checks. See the web route that serves the file. */
private const val LOGIN_CREDS_RELATION = "delegate_permission/common.get_login_creds"

/** What came back from the platform's passkey sheet. */
sealed interface PasskeyResult {
    /** The registration response JSON, ready for GoTrue verbatim. */
    data class Created(val registrationResponseJson: String) : PasskeyResult

    /** Somebody dismissed the sheet. Not a failure to report. */
    data object Dismissed : PasskeyResult

    /**
     * The ceremony failed. [domainNotAssociated] separates the one cause that
     * no amount of retrying fixes, so the card can say so instead of offering
     * advice that cannot work.
     */
    data class Failed(
        val message: String?,
        val domainNotAssociated: Boolean,
    ) : PasskeyResult
}

/**
 * Run the platform's passkey creation sheet.
 *
 * Needs an [Activity] context: Credential Manager presents system UI over the
 * caller, and an application context has nothing to present over.
 */
suspend fun createPasskey(
    context: Context,
    creationOptionsJson: String,
): PasskeyResult {
    val activity = context.findActivity()
        ?: return PasskeyResult.Failed(message = null, domainNotAssociated = false)

    return try {
        val response = CredentialManager.create(activity).createCredential(
            context = activity,
            request = CreatePublicKeyCredentialRequest(requestJson = creationOptionsJson),
        )
        val created = response as? CreatePublicKeyCredentialResponse
            ?: return PasskeyResult.Failed(message = null, domainNotAssociated = false)
        PasskeyResult.Created(created.registrationResponseJson)
    } catch (_: CreateCredentialCancellationException) {
        PasskeyResult.Dismissed
    } catch (cause: CreateCredentialException) {
        PasskeyResult.Failed(
            message = cause.errorMessage?.toString(),
            domainNotAssociated = cause.isDomainAssociationFailure(),
        )
    }
}

/**
 * Whether the failure is Digital Asset Links refusing to associate this app
 * with the relying party.
 *
 * Matched on the message rather than on an exception class because the platform
 * does not give this its own type — it arrives as a generic
 * `CreateCredentialUnknownException` (or a `DomException` subtype) whose text
 * carries the reason. Matching text is fragile, which is why it only ever
 * ADDS a sentence: a miss shows the platform's own message, which is the same
 * thing that would have been shown without this check at all.
 */
private fun CreateCredentialException.isDomainAssociationFailure(): Boolean {
    val haystack = buildString {
        append(type)
        append(' ')
        append(errorMessage ?: "")
    }.lowercase()
    return DOMAIN_FAILURE_MARKERS.any { haystack.contains(it) }
}

private val DOMAIN_FAILURE_MARKERS = listOf(
    // What Play Services says when assetlinks.json does not name this package,
    // is unreachable, or lists a different signing certificate.
    "asset",
    "relying party",
    "not allowed to create a credential",
    "origin",
)

/** The nearest [Activity] up the context chain, or null under an app context. */
private tailrec fun Context.findActivity(): Activity? = when (this) {
    is Activity -> this
    is ContextWrapper -> baseContext.findActivity()
    else -> null
}

File: apps/ios/Loonext/Features/Settings/PasskeyEnrolment.swift
import AuthenticationServices
import Foundation
import UIKit

/// #473 — enrolling a passkey on iOS, through AuthenticationServices.
///
/// ## The case for this being the iOS story rather than a nice extra
///
/// #314 shipped codes from an authenticator app and said in its own words that
/// passkeys suit these users better. It is right, and it is most right here: a
/// tradesperson holds ONE phone, and the authenticator sits on the same screen
/// as the app asking for the six digits it shows. Face ID instead of copying
/// digits between two apps on one screen is the largest single UX win in the
/// issue.
///
/// ## A SECOND FACTOR, NEVER THE PASSWORD (D125)
///
/// The credential lives on this handset and in this iCloud Keychain. Standing
/// alone it would make a lost phone an account reachable only through the
/// recovery codes, which turns the last resort into the primary key.
///
/// ## What travels, and what this file has to build by hand
///
/// Unlike Android — where Credential Manager eats and produces WebAuthn JSON, so
/// the app is a pure courier — Apple's API takes decoded `Data` and returns
/// decoded `Data`. So this file base64url-decodes three fields on the way in and
/// base64url-encodes three on the way out, into exactly the
/// `RegistrationResponseJSON` shape GoTrue expects. Those six conversions are
/// the whole risk surface, and they are why the encoder is a tested pure
/// function rather than six inline expressions.
enum Passkeys {
    /// The relying party every client enrols under.
    ///
    /// The web app enrols with `window.location.hostname`, which is this. A
    /// passkey is scoped to the RP that created it, so a different id here would
    /// mint a credential the web app could never use: one account, two second
    /// factors that each work in exactly one place.
    static let relyingPartyId = "app.loonext.com"

    /// What the caller needs out of the server's creation options.
    struct CreationOptions: Equatable {
        let challenge: Data
        let userId: Data
        let userName: String
    }

    /// Base64url without padding, per the WebAuthn spec's `Base64URLString`.
    ///
    /// Foundation only speaks standard base64, and the difference is not
    /// cosmetic: `+` and `/` appear in roughly half of all 32-byte challenges,
    /// so a naive `base64EncodedString()` produces a value the server rejects
    /// for *some* enrolments and accepts for others. An intermittent failure is
    /// far worse than a total one, which is the reason this is its own function
    /// with its own test rather than a line inside the encoder.
    static func base64UrlEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    /// The inverse. Padding is restored because Foundation's decoder requires
    /// it, and a server is free to send either form.
    static func base64UrlDecode(_ value: String) -> Data? {
        var normalised = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = normalised.count % 4
        if remainder > 0 {
            normalised += String(repeating: "=", count: 4 - remainder)
        }
        return Data(base64Encoded: normalised)
    }

    /// Pull what `ASAuthorizationPlatformPublicKeyCredentialProvider` needs out
    /// of the server's `PublicKeyCredentialCreationOptionsJSON`.
    ///
    /// Returns nil rather than throwing on a shape it does not recognise: the
    /// caller's answer to "the server sent something unexpected" is the same as
    /// its answer to every other enrolment failure, and a thrown error here
    /// would only travel further to be turned back into that same sentence.
    static func creationOptions(fromJson json: String) -> CreationOptions? {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let challengeText = root["challenge"] as? String,
              let challenge = base64UrlDecode(challengeText),
              let user = root["user"] as? [String: Any],
              let userIdText = user["id"] as? String,
              let userId = base64UrlDecode(userIdText)
        else { return nil }

        // `name` is the account handle a passkey manager lists; `displayName` is
        // a human name. Apple shows one string, and the handle is the one that
        // disambiguates two accounts on the same device — which is exactly the
        // case somebody with a personal and a work login is in.
        let name = (user["name"] as? String)
            ?? (user["displayName"] as? String)
            ?? ""
        return CreationOptions(challenge: challenge, userId: userId, userName: name)
    }

    /// Build the `RegistrationResponseJSON` GoTrue's verify step expects.
    ///
    /// The field set matches what the JavaScript SDK produces on a browser
    /// without WebAuthn Level 3 — `id`, `rawId`, `response.attestationObject`,
    /// `response.clientDataJSON`, `type`, `clientExtensionResults` — because
    /// that is the shape the server has always been fed and the one it parses.
    static func registrationResponseJson(
        credentialId: Data,
        clientDataJson: Data,
        attestationObject: Data
    ) -> String? {
        let id = base64UrlEncode(credentialId)
        let payload: [String: Any] = [
            "id": id,
            "rawId": id,
            "type": "public-key",
            "response": [
                "clientDataJSON": base64UrlEncode(clientDataJson),
                "attestationObject": base64UrlEncode(attestationObject),
            ],
            // Present and empty rather than absent: the server reads this key,
            // and an absent one has been the difference between a parse and a
            // 400 on more than one WebAuthn implementation.
            "clientExtensionResults": [String: String](),
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: payload) else {
            return nil
        }
        return String(data: data, encoding: .utf8)
    }

    /// Whether this domain has authorised this app to hold passkeys for it.
    ///
    /// The same probe Android runs, against the file Apple reads: until
    /// `/.well-known/apple-app-site-association` lists this bundle under
    /// `webcredentials`, every ceremony fails — and it fails the way a bug does.
    /// Offering a button that cannot work is worse than not offering it, and
    /// this audience is standing in somebody's driveway when they try it.
    ///
    /// Read from the web app rather than from a build flag because a flag needs
    /// an App Store release to flip, and the file needs a deploy. Passkeys
    /// appear the day they start working, on phones installed months earlier.
    ///
    /// Any failure is `false`: no network, no passkey offer, and the
    /// authenticator app is right there needing nothing.
    static func isDomainAssociated(
        rpId: String = Passkeys.relyingPartyId,
        bundleId: String? = Bundle.main.bundleIdentifier
    ) async -> Bool {
        guard let bundleId,
              let url = URL(string: "https://\(rpId)/.well-known/apple-app-site-association")
        else { return false }
        do {
            let (data, response) = try await URLSession.shared.data(from: url)
            guard (response as? HTTPURLResponse)?.statusCode == 200 else { return false }
            return associationLists(bundleId: bundleId, in: data)
        } catch {
            return false
        }
    }

    /// Whether an apple-app-site-association document names this bundle.
    ///
    /// Suffix-matched on `.<bundleId>` because the entries are Team-ID
    /// prefixed and the team id is not knowable here without reading the
    /// provisioning profile. That is enough for the question being asked —
    /// "has anybody configured this domain for iOS" — and the platform re-checks
    /// the whole association anyway, refusing on any mismatch. This is a UI
    /// gate, not a security one.
    static func associationLists(bundleId: String, in data: Data) -> Bool {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let webcredentials = root["webcredentials"] as? [String: Any],
              let apps = webcredentials["apps"] as? [String]
        else { return false }
        return apps.contains { $0.hasSuffix(".\(bundleId)") }
    }
}

/// Runs the platform's passkey creation sheet and hands back the credential.
///
/// A class with a strong self-reference for the duration: `ASAuthorizationController`
/// holds its delegate weakly, so a flow driven from a free function would be
/// deallocated between `performRequests()` and the callback, and the sheet would
/// simply never answer.
@MainActor
final class PasskeyEnrolment: NSObject {
    /// Held strong — the controller does not, and neither would anything else
    /// once the enclosing function has suspended.
    private var controller: ASAuthorizationController?
    private var retain: PasskeyEnrolment?
    private var continuation: CheckedContinuation<ASAuthorization?, Error>?

    /// Present the sheet.
    ///
    /// Returns nil when the person dismissed it, which is not an error to
    /// report: shouting at somebody who changed their mind teaches them that the
    /// button was dangerous.
    func createCredential(
        rpId: String,
        options: Passkeys.CreationOptions
    ) async throws -> ASAuthorization? {
        let provider = ASAuthorizationPlatformPublicKeyCredentialProvider(
            relyingPartyIdentifier: rpId
        )
        let request = provider.createCredentialRegistrationRequest(
            challenge: options.challenge,
            name: options.userName,
            userID: options.userId
        )
        let controller = ASAuthorizationController(authorizationRequests: [request])
        controller.delegate = self
        controller.presentationContextProvider = self
        self.controller = controller
        retain = self

        return try await withCheckedThrowingContinuation { continuation in
            self.continuation = continuation
            controller.performRequests()
        }
    }

    /// Resume once and only once. The system is not required to call exactly one
    /// of these, and resuming a continuation twice is a crash rather than a bug
    /// report.
    private func finish(_ result: Result<ASAuthorization?, Error>) {
        guard let continuation else { return }
        self.continuation = nil
        controller = nil
        retain = nil
        continuation.resume(with: result)
    }
}

/// The system calls these on the main thread; the `@preconcurrency` conformance's
/// main-actor assumption always holds. Same pattern as `GoogleSignInFlow`.
extension PasskeyEnrolment: @preconcurrency ASAuthorizationControllerDelegate {
    func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithAuthorization authorization: ASAuthorization
    ) {
        finish(.success(authorization))
    }

    func authorizationController(
        controller: ASAuthorizationController,
        didCompleteWithError error: Error
    ) {
        // A cancel is a decision, not a failure.
        if (error as? ASAuthorizationError)?.code == .canceled {
            finish(.success(nil))
            return
        }
        finish(.failure(error))
    }
}

extension PasskeyEnrolment: @preconcurrency ASAuthorizationControllerPresentationContextProviding {
    func presentationAnchor(for controller: ASAuthorizationController) -> ASPresentationAnchor {
        let windows = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
        return windows.first { $0.isKeyWindow } ?? windows.first ?? ASPresentationAnchor()
    }
}
